### PR TITLE
#13332 limit()

### LINKED ISSUE #13332: limit(): AttributeError: 'NoneType' object has no attribute 'rewrite'
```python
import sympy as sp

def stirling(n):
    return sp.sqrt(2 * sp.pi * n) * (n / sp.E)**n

def binom_stirling(n, k):
    return stirling(n) / (stirling(k) * stirling(n - k))

n, k = sp.symbols('n k')

r0 = sp.simplify(binom_stirling(2*(3*k), k))
r1 = sp.simplify(binom_stirling(2*(3*k+1), k))
r = sp.simplify(r0 / r1)

print(sp.limit(r, k, sp.oo))
```

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-135-c7cb354b6b00> in <module>()
     13 r = sp.simplify(r0 / r1)
     14 
---> 15 print(sp.limit(r, k, sp.oo))

~/.local/lib/python3.5/site-packages/sympy/series/limits.py in limit(e, z, z0, dir)
     47     """
     48 
---> 49     return Limit(e, z, z0, dir).doit(deep=False)
     50 
     51 

~/.local/lib/python3.5/site-packages/sympy/series/limits.py in doit(self, **hints)
    178 
    179         try:
--> 180             r = gruntz(e, z, z0, dir)
    181             if r is S.NaN:
    182                 raise PoleError()

~/.local/lib/python3.5/site-packages/sympy/series/gruntz.py in gruntz(e, z, z0, dir)
    662     # It might be nicer to rewrite the exactly to what they were initially,
    663     # but that would take some work to implement.
--> 664     return r.rewrite('intractable', deep=True)

AttributeError: 'NoneType' object has no attribute 'rewrite'
```

Sympy version is 1.1.1.

### PR DESCRIPTION
**Pull request by souravghosh97**
fixes #13332

### CODE REVIEW

**Comment by gxyd** [comment]
You will need to add test case for this. That will be in file `series/tests/test_limits.py`.

**Comment by ylemkimon** [comment]
If this closes the issue you mentioned, it is better to use keywords like `fixes` or `closes` since GitHub [automatically closes the issue](https://help.github.com/articles/closing-issues-using-keywords/) when the PR is merged.

**Comment by ylemkimon** [comment]
Also the test code seems to be from https://github.com/skirpichev/diofant/commit/9d098a7a596df913229ce6b5feee23ab83d12c90.

File: sympy/series/tests/test_limits.py [hunk 1]
```diff
@@ -505,3 +505,8 @@ def test_issue_6599():
 def test_issue_12555():
     assert limit((3**x + 2* x**10) / (x**10 + exp(x)), x, -oo) == 2
     assert limit((3**x + 2* x**10) / (x**10 + exp(x)), x, oo) == oo
+
+
+def test_sympyissue_13332():
```
Comment (ylemkimon) [inline sympy/series/tests/test_limits.py]: Why `test_sympyissue_13332`, rather than `test_issue_13332`?

File: sympy/series/tests/test_limits.py [hunk 2]
```diff
@@ -505,3 +505,8 @@ def test_issue_6599():
...
+    assert limit(sqrt(30)*5**(-5*n - 1)*(46656*n)**n*(5*n + 2)**(5*n + 5*S.Half) *
+                 (6*n + 2)**(-6*n - 5*S.Half), n, oo) == Rational(25, 36)
```
Comment (jksuom) [inline sympy/series/tests/test_limits.py]: A newline is expected at end of line.

File: sympy/series/gruntz.py [hunk 3]
```diff
@@ -409,7 +410,7 @@ def sign(e, x):
 @debug
 @timeit
 @cacheit
-def limitinf(e, x):
+def limitinf(e, x,simp=False):
```
Comment (andrew-boyarshin) [inline sympy/series/gruntz.py]: Feels like a space char is missing there.

**Comment by czgdp1807** [comment]
@souravghosh97 Are you still working on it?

**Comment by souravghosh97** [comment]
No.

**Comment by czgdp1807** [comment]
@souravghosh97 No problem. Thanks for the contribution.
@ylemkimon @jksuom It looks like this PR successfully fixes the issue.
Can we continue this PR and make a new one to complete it?
```python
Python 3.6.8 (default, Aug 20 2019, 17:12:48) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sympy as sp
>>> def binom_stirling(n, k):
...     return stirling(n) / (stirling(k) * stirling(n - k))
... 
>>> def stirling(n):
...     return sp.sqrt(2 * sp.pi * n) * (n / sp.E)**n
... 
>>> n, k = sp.symbols('n k')
>>> r0 = sp.simplify(binom_stirling(2*(3*k), k))
>>> r1 = sp.simplify(binom_stirling(2*(3*k+1), k))
>>> r = sp.simplify(r0 / r1)
>>> sp.limit(r, k, sp.oo)
25/36 
```

**Comment by jksuom** [comment]
> Can we continue this PR and make a new one to complete it?

Yes, simplifying the leading term of a series expansion seems a good idea to me. It will resolve some cases where the original function cannot be sufficiently simplified.

I would make two changes: combine the try blocks of `gruntz` into a single block, and make `limitinf` raise ValueError instead of returning None if the sign of the leading exponent e0 cannot be evaluated. (Then None need not be tested.) This is the diff, including some minor edits:
```
diff --git a/sympy/series/gruntz.py b/sympy/series/gruntz.py
index e9dc986..80867ae 100644
--- a/sympy/series/gruntz.py
+++ b/sympy/series/gruntz.py
@@ -122,7 +122,6 @@
 from sympy.functions import log, exp
 from sympy.series.order import Order
 from sympy.simplify.powsimp import powsimp, powdenest
-from sympy.simplify.simplify import simplify
 from sympy import cacheit
 
 from sympy.core.compatibility import reduce
@@ -410,8 +409,13 @@ def sign(e, x):
 @debug
 @timeit
 @cacheit
-def limitinf(e, x, simp=False):
-    """Limit e(x) for x-> oo"""
+def limitinf(e, x, leadsimp=False):
+    """Limit e(x) for x-> oo.
+
+    If ``leadsimp`` is True, an attempt is made to simplify the leading
+    term of the series expansion of ``e``. That may succeed even if
+    ``e`` cannot be simplified.
+    """
     # rewrite e in terms of tractable functions only
     e = e.rewrite('tractable', deep=True)
 
@@ -439,9 +443,11 @@ def limitinf(e, x, simp=False):
             raise ValueError("Leading term should not be 0")
         return s*oo
     elif sig == 0:
-        if simp:
-            c0 = simplify(c0)
-        return limitinf(c0, x, simp)  # e0=0: lim f = lim c0
+        if leadsimp:
+            c0 = c0.simplify()
+        return limitinf(c0, x, leadsimp)  # e0=0: lim f = lim c0
+    else:
+        raise ValueError("{} could not be evaluated".format(sig))
 
 
 def moveup2(s, x):
@@ -648,19 +654,9 @@ def gruntz(e, z, z0, dir="+"):
     # convert all limits to the limit z->oo; sign of z is handled in limitinf
     r = None
     if z0 == oo:
-        try:
-            r = limitinf(e, z)
-        except ValueError:
-            r = limitinf(e, z, simp=True)
-        if r is None:
-            r = limitinf(e, z, simp=True)
+        e0 = e
     elif z0 == -oo:
-        try:
-            r = limitinf(e.subs(z, -z), z)
-        except ValueError:
-            r = limitinf(e.subs(z, -z), z, simp=True)
-        if r is None:
-            r = limitinf(e.subs(z, -z), z, simp=True)
+        e0 = e.subs(z, -z)
     else:
         if str(dir) == "-":
             e0 = e.subs(z, z0 - 1/z)
@@ -668,7 +664,11 @@ def gruntz(e, z, z0, dir="+"):
             e0 = e.subs(z, z0 + 1/z)
         else:
             raise NotImplementedError("dir must be '+' or '-'")
+
+    try:
         r = limitinf(e0, z)
+    except ValueError:
+        r = limitinf(e0, z, leadsimp=True)
 
     # This is a bit of a heuristic for nice results... we always rewrite
     # tractable functions in terms of familiar intractable ones.
```
[diff.txt](https://github.com/sympy/sympy/files/3690267/diff.txt)

**Comment by jksuom** [comment]
I made some effort towards replacing `simplify` with a more light-weight method but could not find one general enough. It looks like the bulk of time will taken by other functions of the gruntz module anyway.

**Pull request closed without merging**